File: frontend/static/learner.js
const app = document.getElementById('app');
const esc = value => String(value ?? '').replace(/[&<>"']/g, c => ({'&':'&amp;','<':'&lt;','>':'&gt;','"':'&quot;',"'":'&#39;'}[c]));
const date = value => value ? new Date(value).toLocaleString() : 'No deadline';
const number = value => Number(value || 0).toLocaleString();

async function request(url, options) {
  const response = await fetch(url, {...options, headers: {'Accept':'application/json', ...(options && options.headers)}});
  const data = await response.json();
  if (!response.ok) throw new Error(data.error || 'Request failed');
  return data;
}

function taskMarkup(task, vm, index, readOnly) {
  const hints = task.hints.map(h => `<div class="hint" data-hint="${h.index}">${h.revealed ? esc(h.text) : `<button data-reveal="${h.index}">Reveal hint ${h.index + 1}</button>`}</div>`).join('');
  const debrief = task.debrief ? `<section class="debrief"><h4>Debrief</h4><p><strong>Root cause:</strong> ${esc(task.debrief.root_cause)}</p><p><strong>Remediation:</strong> ${esc(task.debrief.remediation)}</p><p><strong>ATT&amp;CK:</strong> ${esc(task.debrief.attack_mapping)}</p>${task.references.map(r => `<p><a href="${esc(r)}" rel="noreferrer">${esc(r)}</a></p>`).join('')}</section>` : '';
  return `<article class="task" data-vm="${vm.id}" data-module="${esc(task.id)}"><div class="task-index">${String(index + 1).padStart(2,'0')}</div><div class="task-main"><h3>${esc(task.name)}</h3><p>${esc(task.description)}</p><div class="tags"><span class="status status-${esc(task.status)}">${esc(task.status)}</span><span class="tag">${esc(task.difficulty)}</span><span class="tag">${task.estimated_minutes} minutes</span><span class="tag">${number(task.points)} points</span></div>${task.prerequisites.length ? `<p><strong>Prerequisites:</strong> ${task.prerequisites.map(esc).join(', ')}</p>` : ''}${task.learning_objectives.length ? `<h4>Learning objectives</h4><ul class="objectives">${task.learning_objectives.map(o => `<li>${esc(o)}</li>`).join('')}</ul>` : ''}<div class="hint-list">${hints}</div>${debrief}</div><aside class="task-side"><button class="button" data-verify ${readOnly ? 'disabled' : ''}>Verify fix</button><small>${task.last_verified_at ? `Last checked ${date(task.last_verified_at)}` : 'Not yet checked'}</small></aside></article>`;
}

function dashboard(data) {
  app.innerHTML = `<section class="event-head"><div class="event-copy"><p class="eyebrow">${esc(data.team.name)} · ${esc(data.event.status)}</p><h1>${esc(data.event.name)}</h1>${data.event.description ? `<p class="event-description">${esc(data.event.description)}</p>` : ''}</div><div class="event-meta"><span class="label">Deadline</span><strong>${esc(date(data.event.ends_at))}</strong></div></section><section class="metric-grid"><div class="metric metric-primary"><span class="label">Team score</span><span class="metric-value">${number(data.score.total)}</span></div><div class="metric"><span class="label">Completed</span><span class="metric-value">${number(data.score.completed)}</span></div><div class="metric"><span class="label">Progress</span><span class="metric-value">${number(data.score.completion_percentage)}%</span></div><div class="metric"><span class="label">Regressions</span><span class="metric-value">${number(data.regressions)}</span></div></section><div class="action-row"><div>${data.read_only ? '<span class="warning">This event is read-only. Historical results remain available.</span>' : data.regressions ? '<span class="warning">A completed task has regressed and no longer contributes points.</span>' : 'Connect to an assigned VM and verify each root-cause fix.'}</div><button class="button button-secondary" id="show-access">Show team access</button></div>${data.vms.map(vm => `<section class="vm-section"><header class="vm-head"><div><p class="label">Assigned target</p><h2>${esc(vm.hostname || `VM ${vm.id}`)}</h2></div><code class="connection">${esc(vm.connection_command)}</code></header>${vm.modules.map((task,i) => taskMarkup(task,vm,i,data.read_only)).join('') || '<p>No learner tasks are assigned to this target.</p>'}</section>`).join('')}${data.recent_activity.length ? `<section class="vm-section"><header class="vm-head"><div><p class="label">Verification history</p><h2>Recent attempts</h2></div></header><div class="table-wrap"><table><thead><tr><th>Time</th><th>Trigger</th><th>Result</th><th>Summary</th></tr></thead><tbody>${data.recent_activity.map(item=>`<tr><td>${esc(date(item.created_at))}</td><td>${esc(item.trigger)}</td><td>${esc(item.result)}</td><td>${esc(item.summary)}</td></tr>`).join('')}</tbody></table></div></section>` : ''}`;
  bindDashboard();
}

function bindDashboard() {
  document.getElementById('show-access').addEventListener('click', async () => {
    const dialog = document.getElementById('access-dialog'), content = document.getElementById('access-content');
    content.innerHTML = '<p>Loading credentials…</p>'; dialog.showModal();
    try { const data = await request('/api/me/team-access'); content.innerHTML = `<h2>Team access</h2><p>Save the private key with mode <code>0600</code>. Use the separate password when sudo prompts.</p><h3>Private key</h3><pre>${esc(data.private_key)}</pre><h3>Sudo password</h3><pre>${esc(data.sudo_password)}</pre><h3>Connections</h3>${data.connections.map(c => `<pre>${esc(c.command)}</pre>`).join('')}`; } catch (error) { content.innerHTML = `<p class="warning">${esc(error.message)}</p>`; }
  });
  app.querySelectorAll('[data-reveal]').forEach(button => button.addEventListener('click', async () => {
    const task = button.closest('.task');
    try { const data = await request(`/api/vms/${task.dataset.vm}/modules/${task.dataset.module}/hints/${button.dataset.reveal}/reveal`, {method:'POST'}); button.parentElement.textContent = data.text; } catch (error) { button.parentElement.textContent = error.message; }
  }));
  app.querySelectorAll('[data-verify]').forEach(button => button.addEventListener('click', async () => {
    const task = button.closest('.task'); button.disabled = true; button.textContent = 'Checking…';
    try { const data = await request(`/api/vms/${task.dataset.vm}/modules/${task.dataset.module}/verify`, {method:'POST'}); button.textContent = data.result === 'pass' ? 'Verified' : data.summary; setTimeout(loadDashboard, 900); } catch (error) { button.textContent = error.message; button.disabled = false; }
  }));
}

async function loadDashboard() { try { dashboard(await request('/api/me/training')); } catch (error) { app.innerHTML = `<p class="warning loading">${esc(error.message)}</p>`; } }
async function loadScoreboard() { try { const data = await request('/api/scoreboard'); app.innerHTML = `<section class="event-head"><div class="event-copy"><p class="eyebrow">${esc(data.event.status)}</p><h1>Scoreboard</h1><p class="event-description">${esc(data.event.name)}</p></div></section><div class="table-wrap"><table><thead><tr><th>Rank</th><th>Team</th><th>Defensive</th><th>Reactive</th><th>Total</th><th>Completion</th><th>Red pressure</th></tr></thead><tbody>${data.teams.map(t => `<tr data-current="${t.team_id === data.current_team_id}"><td>${t.rank}</td><td><strong>${esc(t.team_name)}</strong></td><td>${number(t.blue_defensive)}</td><td>${number(t.blue_reactive)}</td><td>${number(t.total_score)}</td><td>${number(t.completion_percentage)}%</td><td>${number(t.red_team_pressure)}</td></tr>`).join('')}</tbody></table></div>`; } catch (error) { app.innerHTML = `<p class="warning loading">${esc(error.message)}</p>`; } }

document.body.dataset.view === 'scoreboard' ? loadScoreboard() : loadDashboard();

File: api/routes/learner.py
from __future__ import annotations

import json
import os

from fastapi import APIRouter, Depends, Request
from fastapi.responses import JSONResponse
from fastapi.responses import PlainTextResponse
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from api.database import get_db
from api.models import AdminAudit, HintReveal, Team, User, VerificationAttempt, VM, VMGoal, VMModule
from api.services.authorization import learner_assignment, participant
from api.services.gamenet import site_dns_zone, vm_dns_name
from api.services.secrets import decrypt_secret
from api.services.verification import verify_assignment
from api.services.verifier_account import scoring_enabled_vm_ids
from builder.module_loader import load_all_modules

router = APIRouter(prefix="/api", tags=["learner"])


@router.get("/me/gamenet")
async def gamenet_status(request: Request, db: Session = Depends(get_db)):
    user, error = _denied(request, db)
    if error:
        return error
    gateway = user.team.vpn_gateway
    credential = user.vpn_credential
    return {
        "ready": bool(gateway and gateway.status == "active" and credential and credential.status == "active"),
        "status": gateway.status if gateway else "not_provisioned",
        "configuration_url": "/api/me/gamenet/config" if gateway and gateway.status == "active" else None,
        "resolver_address": gateway.vpn_address if gateway else None,
        "site_dns_zones": [
            {"site_key": site.key, "zone": site_dns_zone(site)}
            for site in sorted(user.team.sites, key=lambda item: item.order)
        ],
    }


@router.get("/me/gamenet/config")
async def download_gamenet_config(request: Request, db: Session = Depends(get_db)):
    user, error = _denied(request, db)
    if error:
        return error
    from api.services.gamenet import render_user_config
    try:
        config = render_user_config(db, user)
    except RuntimeError:
        return JSONResponse({"error": "GameNet VPN is not ready"}, status_code=503,
                            headers={"Cache-Control": "no-store"})
    db.add(AdminAudit(actor_id=user.id, target_user_id=user.id, action="gamenet_config_downloaded",
                      metadata_json=json.dumps({"team_id": user.team_id})))
    db.commit()
    return PlainTextResponse(config, media_type="application/x-wireguard-profile",
                             headers={"Cache-Control": "no-store, private", "Pragma": "no-cache",
                                      "Content-Disposition": 'attachment; filename="gamenet.conf"'})


def enabled() -> bool:
    return os.environ.get("LEARNER_TRAINING_ENABLED", "false").lower() in {"1", "true", "yes"}


def _denied(request: Request, db: Session):
    if not enabled():
        return None, JSONResponse({"error": "learner training is not enabled"}, status_code=404)
    user = participant(request, db)
    if not user:
        return None, JSONResponse({"error": "forbidden"}, status_code=403)
    return user, None


def _library():
    return {module.id: module for module in load_all_modules()
            if not module.disabled and module.stage == "preapplied"
            and module.type in {"vulnerability", "hardening", "payload"}}


def _module_payload(assignment: VMModule, definition, revealed: set[int]) -> dict:
    unlocked = assignment.first_completed_at is not None
    return {
        "id": definition.id,
        "assignment_id": assignment.id,
        "name": definition.name,
        "description": definition.description,
        "learning_objectives": definition.learning_objectives,
        "difficulty": definition.difficulty,
        "points": definition.points,
        "prerequisites": definition.prerequisites,
        "estimated_minutes": definition.estimated_minutes,
        "status": assignment.status,
        "last_verified_at": assignment.last_verified_at.isoformat() if assignment.last_verified_at else None,
        "hints": [{"index": index, "revealed": index in revealed,
                   "text": hint if index in revealed else None} for index, hint in enumerate(definition.hints)],
        "debrief": definition.debrief if unlocked else None,
        "references": [reference.as_dict() for reference in definition.references] if unlocked else [],
    }


def _score(db: Session, user: User, assignments: list[VMModule]) -> dict:
    goals = db.query(VMGoal).join(VM).filter(VM.team_id == user.team_id, VM.event_id == user.event_id).all()
    enabled_ids = scoring_enabled_vm_ids(db, {item.vm_id for item in assignments} | {goal.vm_id for goal in goals})
    defensive = sum(item.points for item in assignments if item.vm_id in enabled_ids and item.status == "completed")
    reactive = sum(goal.defend_points * goal.defend_count for goal in goals if goal.vm_id in enabled_ids)
    total = len(assignments)
    completed = sum(item.vm_id in enabled_ids and item.status == "completed" for item in assignments)
    return {"blue_defensive": defensive, "blue_reactive": reactive,
            "total": defensive + reactive, "completed": completed, "assigned": total,
            "completion_percentage": round(completed * 100 / total, 1) if total else 0}


@router.get("/me/training")
async def my_training(request: Request, db: Session = Depends(get_db)):
    user, error = _denied(request, db)
    if error:
        return error
    definitions = _library()
    vms = db.query(VM).filter(VM.team_id == user.team_id, VM.event_id == user.event_id).order_by(VM.hostname).all()
    vm_ids = [vm.id for vm in vms]
    assignments = db.query(VMModule).filter(
        VMModule.vm_id.in_(vm_ids), VMModule.stage == "preapplied",
        VMModule.module_type.in_(("vulnerability", "hardening", "payload")),
    ).all() if vm_ids else []
    visible = [item for item in assignments if item.module_id in definitions]
    reveal_rows = db.query(HintReveal).filter(
        HintReveal.user_id == user.id,
        HintReveal.module_assignment_id.in_([item.id for item in visible]),
    ).all() if visible else []
    reveals: dict[int, set[int]] = {}
    for row in reveal_rows:
        reveals.setdefault(row.module_assignment_id, set()).add(row.hint_index)
    attempts = db.query(VerificationAttempt).join(VMModule).filter(
        VMModule.vm_id.in_(vm_ids)
    ).order_by(VerificationAttempt.created_at.desc()).limit(20).all() if vm_ids else []
    by_vm: dict[int, list[VMModule]] = {vm.id: [] for vm in vms}
    for item in visible:
        by_vm[item.vm_id].append(item)
    event = user.event
    return {
        "event": {"id": event.id, "name": event.name, "status": event.status,
                  "description": event.description, "ends_at": event.ends_at.isoformat() if event.ends_at else None},
        "team": {"id": user.team.id, "name": user.team.name},
        "vms": [{"id": vm.id, "hostname": vm.hostname, "address": vm.private_ip or vm.ip_address,
                 "dns_name": vm_dns_name(vm),
                 "ssh_port": vm.ssh_port or 22, "status": vm.status,
                 "connection_command": f"ssh -p {vm.ssh_port or 22} ctf-trainee@{vm_dns_name(vm) or vm.private_ip or vm.ip_address}",
                 "modules": [_module_payload(item, definitions[item.module_id], reveals.get(item.id, set()))
                             for item in sorted(by_vm[vm.id], key=lambda value: value.id)]} for vm in vms],
        "score": _score(db, user, visible),
        "regressions": sum(item.status == "regressed" for item in visible),
        "recent_activity": [{"module_assignment_id": attempt.module_assignment_id,
                             "result": attempt.result, "summary": attempt.safe_summary,
                             "trigger": attempt.trigger_type, "created_at": attempt.created_at.isoformat()}
                            for attempt in attempts],
        "read_only": event.status != "open",
    }


@router.get("/me/team-access")
async def team_access(request: Request, db: Session = Depends(get_db)):
    user, error = _denied(request, db)
    if error:
        return error
    credential = user.team.training_credential
    if not credential or credential.status != "active":
        return JSONResponse({"error": "team credentials are not ready"}, status_code=503,
                            headers={"Cache-Control": "no-store"})
    vms = db.query(VM).filter(VM.team_id == user.team_id, VM.event_id == user.event_id).all()
    db.add(AdminAudit(actor_id=user.id, target_user_id=user.id, action="team_credential_revealed",
                      metadata_json=json.dumps({"team_id": user.team_id})))
    db.commit()
    return JSONResponse({
        "username": credential.username,
        "private_key": decrypt_secret(credential.private_key_encrypted),
        "public_key": credential.public_key,
        "sudo_password": decrypt_secret(credential.sudo_password_encrypted),
        "connections": [{"vm_id": vm.id, "hostname": vm.hostname,
                         "dns_name": vm_dns_name(vm), "address": vm.private_ip or vm.ip_address,
                         "command": f"ssh -i ./ctf-team-key -p {vm.ssh_port or 22} {credential.username}@{vm_dns_name(vm) or vm.private_ip or vm.ip_address}"}
                        for vm in vms],
    }, headers={"Cache-Control": "no-store, private", "Pragma": "no-cache"})


@router.post("/vms/{vm_id}/modules/{module_id}/verify")
async def verify(vm_id: int, module_id: str, request: Request, db: Session = Depends(get_db)):
    user, error = _denied(request, db)
    if error:
        return error
    if user.event.status != "open":
        return JSONResponse({"error": "event is read-only"}, status_code=409)
    assignment = learner_assignment(db, user, vm_id, module_id)
    definition = _library().get(module_id)
    if not assignment or not definition:
        return JSONResponse({"error": "module not found"}, status_code=404)
    result = await verify_assignment(db, assignment, definition.verification, "learner", user)
    return {"result": result.result, "summary": result.summary, "error_code": result.error_code,
            "status": assignment.status, "score": _score(db, user, db.query(VMModule).join(VM).filter(
                VM.team_id == user.team_id, VM.event_id == user.event_id, VMModule.stage == "preapplied").all())}


@router.post("/vms/{vm_id}/modules/{module_id}/hints/{index}/reveal")
async def reveal_hint(vm_id: int, module_id: str, index: int, request: Request, db: Session = Depends(get_db)):
    user, error = _denied(request, db)
    if error:
        return error
    if user.event.status != "open":
        return JSONResponse({"error": "event is read-only"}, status_code=409)
    assignment = learner_assignment(db, user, vm_id, module_id)
    definition = _library().get(module_id)
    if not assignment or not definition:
        return JSONResponse({"error": "module not found"}, status_code=404)
    if index < 0 or index >= len(definition.hints):
        return JSONResponse({"error": "hint not found"}, status_code=404)
    existing = db.query(HintReveal).filter_by(user_id=user.id, module_assignment_id=assignment.id, hint_index=index).first()
    if index > 0 and not db.query(HintReveal).filter_by(
        user_id=user.id, module_assignment_id=assignment.id, hint_index=index - 1
    ).first():
        return JSONResponse({"error": "reveal the previous hint first"}, status_code=409)
    if not existing:
        db.add(HintReveal(user_id=user.id, module_assignment_id=assignment.id, hint_index=index))
        try:
            db.commit()
        except IntegrityError:
            db.rollback()
    return {"index": index, "text": definition.hints[index], "points_penalty": 0}


@router.get("/scoreboard")
async def scoreboard(request: Request, db: Session = Depends(get_db)):
    user, error = _denied(request, db)
    if error:
        return error
    teams = db.query(Team).filter(Team.event_id == user.event_id).all()
    rows = []
    for team in teams:
        modules = db.query(VMModule).join(VM).filter(VM.team_id == team.id, VM.event_id == user.event_id,
            VMModule.stage == "preapplied").all()
        goals = db.query(VMGoal).join(VM).filter(VM.team_id == team.id, VM.event_id == user.event_id).all()
        enabled_ids = scoring_enabled_vm_ids(db, {module.vm_id for module in modules} | {goal.vm_id for goal in goals})
        defensive = sum(module.points for module in modules if module.vm_id in enabled_ids and module.status == "completed")
        reactive = sum(goal.defend_points * goal.defend_count for goal in goals if goal.vm_id in enabled_ids)
        red = sum(goal.red_points * goal.achievement_count for goal in goals if goal.vm_id in enabled_ids)
        completed = sum(module.vm_id in enabled_ids and module.status == "completed" for module in modules)
        rows.append({"team_id": team.id, "team_name": team.name, "blue_defensive": defensive,
                     "blue_reactive": reactive, "total_score": defensive + reactive,
                     "completion_percentage": round(completed * 100 / len(modules), 1) if modules else 0,
                     "red_team_pressure": red})
    rows.sort(key=lambda row: (-row["total_score"], -row["completion_percentage"], row["team_name"].lower()))
    for rank, row in enumerate(rows, 1):
        row["rank"] = rank
    return {"event": {"id": user.event.id, "name": user.event.name, "status": user.event.status},
            "current_team_id": user.team_id, "teams": rows}
